### PR TITLE
Rework error handling

### PR DESCRIPTION
**Pull request by erlend-aasland**
- Introduce ClinicError and ClinicWarning
- fail() raises ClinicError
- warn() uses warnings.warn(msg, ClinicWarning)
- the CLI runs main(), catches ClinicError, formats the error message prints to stderr and exits with an error
- adapt the test suite to work with ClinicError

### CODE REVIEW

**Comment by erlend-aasland** [comment]
@AlexWaygood, what do you think of this approach?

**Comment by erlend-aasland** [comment]
Can be broken up like this:

- PR 1: split out `create_cli()`
- PR 2: add and use ClinicWarning
- PR 3: add and use ClinicError (including test adjustments)

**Comment by erlend-aasland** [comment]
(BTW, since I touched pretty much every test that checks failure of some kind, I just normalised variable naming for them all)

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -136,6 +138,24 @@ def text_accumulator() -> TextAccumulator:
...
+
+class ClinicError(Exception):
+    def __init__(
+        self,
+        message: str,
+        /,
+        *,
+        lineno: int | None = None,
+        filename: str | None = None
+    ):
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:150]:
```suggestion
    ) -> None:
```

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -136,6 +138,24 @@ def text_accumulator() -> TextAccumulator:
...
+        super().__init__(message)
+        self.lineno = lineno
+        self.filename = filename
+
+
+class ClinicWarning(Warning): ...
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:156]: Style nit: the convention (as far as I understand it) is that stub files, and other places where you'd never _expect_ to find a body (abstract methods, methods on protocols, methods decorated with `@overload`) should use `...` for empty bodies. But for places where you'd usually _expect_ to find some kind of a body (any other kind of a function in a non-stub file; any class definition in a non-stub file) should generally be explicit about the empty body by using `pass`:

```suggestion
class ClinicWarning(Warning): pass
```

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -5620,9 +5636,10 @@ def state_terminal(self, line: str | None) -> None:
 clinic = None
 
 
-def main(argv: list[str]) -> None:
-    import sys
-    import argparse
+class CLIError(Exception): ...
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:5639]:
```suggestion
class CLIError(Exception): pass
```

File: Tools/clinic/clinic.py [hunk 4]
```diff
@@ -5751,5 +5757,19 @@ def main(argv: list[str]) -> None:
 
 
 if __name__ == "__main__":
-    main(sys.argv[1:])
-    sys.exit(0)
+    cli = create_cli()
+    try:
+        args = cli.parse_args()
+        main(args)
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:5763]: Just for symmetry with the other branches, more than anything else:

```suggestion
        main(args)
        sys.exit(0)
```

**Review by AlexWaygood — COMMENTED** [review]
Looks great overall. A few comments below, but nothing major:

File: Lib/test/test_clinic.py [hunk 5]
```diff
@@ -140,51 +138,43 @@ def test_mangled_marker_line(self):
             [clinic start generated code]*/
             /*[clinic end generated code: foo]*/
         """
-        msg = (
-            'Error in file "test.c" on line 3:\n'
-            "Mangled Argument Clinic marker line: '/*[clinic end generated code: foo]*/'\n"
+        err = (
```
[the remark this answers is not in the document]
Comment (erlend-aasland) [reply Lib/test/test_clinic.py:141]: Yes, of course!

**Comment by AlexWaygood** [comment]
I have a few more thoughts after you've taken a look at my first round of review ;)

But again, nothing major!

**Comment by erlend-aasland** [comment]
Since you agree with the overall approach, let's take this to the cpython repo in smaller steps :)

**Comment by AlexWaygood** [comment]
Yep. One more thing I would say is that it would be nice to restructure the top-level `if __name__ == "__main__"` block like this:

```py
def main(argv: list[str] | None = None) -> None:
    argv = argv or sys.argv
    cli = create_cli()
    try:
        args = cli.parse_args(argv)
        run_clinic(args)
        sys.exit(0)
    except CLIError as exc:
        if msg := str(exc):
            sys.stderr.write(f"Usage error: {msg}\n")
        cli.print_usage()
        sys.exit(1)
    except ClinicError as exc:
        msg = textwrap.dedent(f"""\
            Error in file {exc.filename!r} on line {exc.lineno}:
            {exc}
        """)
        sys.stderr.write(str(exc))
        sys.exit(1)

if __name__ == "__main__":
    main()
```

(`ArgumentParser.parse_args()` takes an optional parameter -- if provided, it will use that parameter instead of `sys.argv`: 

If we did that, it would be very easy to do end-to-end tests of the CLI _without_ using subprocesses. (The tests got a fair bit slower for me locally on Windows since we started using subprocesses -- they have a lot of overhead on Windows, sadly 😞). When the CLI is actually run, it'll use `sys.argv`, but in tests, we can pass arbitrary arguments to the `main()` function to accurately mock an real run of the CLI.

**Comment by erlend-aasland** [comment]
> If we did that, it would be very easy to do end-to-end tests of the CLI _without_ using subprocesses. (The tests got a fair bit slower for me locally on Windows since we started using subprocesses -- they have a lot of overhead on Windows, sadly 😞). When the CLI is actually run, `cli.parse_args()` will use `sys.argv`, but in tests, we can pass arbitrary arguments to the `main()` function to accurately mock a real run of the CLI.

That's a very good idea. Reducing CI (and local dev workflow) time is a very good thing.

**Comment by erlend-aasland** [comment]
https://github.com/python/cpython/pull/107551

**Pull request closed without merging**